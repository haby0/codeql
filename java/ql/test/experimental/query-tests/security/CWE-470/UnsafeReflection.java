import java.util.HashSet;
import javax.servlet.http.HttpServletRequest;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.GetMapping;

@Controller
public class UnsafeReflection {

    @GetMapping(value = "uf1")
    public void bad1(HttpServletRequest request) {
        String className = request.getParameter("className");
        String parameterValue = request.getParameter("parameterValue");
        try {
            Class clazz = Class.forName(className);
            Object object = clazz.getDeclaredConstructors()[0].newInstance(parameterValue); //bad
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @GetMapping(value = "uf2")
    public void bad2(HttpServletRequest request) {
        String className = request.getParameter("className");
        String parameterValue = request.getParameter("parameterValue");
        try {
            ClassLoader classLoader = ClassLoader.getSystemClassLoader();
            Class clazz = classLoader.loadClass(className);
            Object object = clazz.newInstance();
            clazz.getDeclaredMethods()[0].invoke(object, parameterValue); //bad
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @GetMapping(value = "uf3")
    public void bad3(HttpServletRequest request) {
        String className = request.getParameter("className");
        String parameterValue = request.getParameter("parameterValue");
        try {
            ClassLoader classLoader = ClassLoader.getSystemClassLoader();
            Class clazz = classLoader.loadClass(className);
            Object object = clazz.newInstance(); //bad
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @GetMapping(value = "uf4")
    public void good1(HttpServletRequest request) throws Exception {
        HashSet<String> hashSet = new HashSet<>();
        hashSet.add("com.example.test1");
        hashSet.add("com.example.test2");
        String className = request.getParameter("className");
        String parameterValue = request.getParameter("parameterValue");
        if (!hashSet.contains(className)){ 
            throw new Exception("Class not valid: "  + className);
        }
        try {
            Class clazz = Class.forName(className);
            Object object = clazz.getDeclaredConstructors()[0].newInstance(parameterValue); //good
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @GetMapping(value = "uf5")
    public void good2(HttpServletRequest request) throws Exception {
        String className = request.getParameter("className");
        String parameterValue = request.getParameter("parameterValue");
        if (!"com.example.test1".equals(className)){
            throw new Exception("Class not valid: "  + className);
        }
        try {
            Class clazz = Class.forName(className);
            Object object = clazz.getDeclaredConstructors()[0].newInstance(parameterValue); //good
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @GetMapping(value = "uf6")
    public void good3(HttpServletRequest request) throws Exception {
        String className = request.getParameter("className");
        String parameterValue = request.getParameter("parameterValue");
        if (!className.equals("com.example.test1")){ //good
            throw new Exception("Class not valid: "  + className);
        }
        try {
            Class clazz = Class.forName(className);
            Object object = clazz.getDeclaredConstructors()[0].newInstance(parameterValue); //good
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}
