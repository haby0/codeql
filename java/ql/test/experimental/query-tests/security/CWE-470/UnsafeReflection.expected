edges
| UnsafeReflection.java:11:28:11:60 | getParameter(...) : String | UnsafeReflection.java:15:29:15:62 | ...[...] |
| UnsafeReflection.java:23:28:23:60 | getParameter(...) : String | UnsafeReflection.java:28:29:28:33 | clazz |
| UnsafeReflection.java:23:28:23:60 | getParameter(...) : String | UnsafeReflection.java:29:13:29:41 | ...[...] |
| UnsafeReflection.java:37:28:37:60 | getParameter(...) : String | UnsafeReflection.java:42:29:42:33 | clazz |
nodes
| UnsafeReflection.java:11:28:11:60 | getParameter(...) : String | semmle.label | getParameter(...) : String |
| UnsafeReflection.java:15:29:15:62 | ...[...] | semmle.label | ...[...] |
| UnsafeReflection.java:23:28:23:60 | getParameter(...) : String | semmle.label | getParameter(...) : String |
| UnsafeReflection.java:28:29:28:33 | clazz | semmle.label | clazz |
| UnsafeReflection.java:29:13:29:41 | ...[...] | semmle.label | ...[...] |
| UnsafeReflection.java:37:28:37:60 | getParameter(...) : String | semmle.label | getParameter(...) : String |
| UnsafeReflection.java:42:29:42:33 | clazz | semmle.label | clazz |
#select
| UnsafeReflection.java:15:29:15:62 | ...[...] | UnsafeReflection.java:11:28:11:60 | getParameter(...) : String | UnsafeReflection.java:15:29:15:62 | ...[...] | Unsafe reflection of $@. | UnsafeReflection.java:11:28:11:60 | getParameter(...) | user input |
| UnsafeReflection.java:28:29:28:33 | clazz | UnsafeReflection.java:23:28:23:60 | getParameter(...) : String | UnsafeReflection.java:28:29:28:33 | clazz | Unsafe reflection of $@. | UnsafeReflection.java:23:28:23:60 | getParameter(...) | user input |
| UnsafeReflection.java:29:13:29:41 | ...[...] | UnsafeReflection.java:23:28:23:60 | getParameter(...) : String | UnsafeReflection.java:29:13:29:41 | ...[...] | Unsafe reflection of $@. | UnsafeReflection.java:23:28:23:60 | getParameter(...) | user input |
| UnsafeReflection.java:42:29:42:33 | clazz | UnsafeReflection.java:37:28:37:60 | getParameter(...) : String | UnsafeReflection.java:42:29:42:33 | clazz | Unsafe reflection of $@. | UnsafeReflection.java:37:28:37:60 | getParameter(...) | user input |
